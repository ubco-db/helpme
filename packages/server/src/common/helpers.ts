import { Request } from 'express';

export function getCookie(req: Request, cookieName: string): string {
  const cookieHeader = req.headers.cookie;
  if (!cookieHeader) {
    return null;
  }

  const cookies: string[] = cookieHeader.split(';');
  for (const cookie of cookies) {
    const [name, value] = cookie.split('=');
    if (name.trim() === cookieName) {
      return value;
    }
  }
}
