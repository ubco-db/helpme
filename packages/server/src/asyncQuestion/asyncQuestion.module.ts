import { Module } from '@nestjs/common';
import { NotificationModule } from '../notification/notification.module';
import { asyncQuestionController } from './asyncQuestion.controller';
import { asyncQuestionService } from './asyncQuestion.service';
import { MailModule, MailTestingModule } from 'mail/mail.module';
import { QuestionService } from '../question/question.service';

@Module({
  controllers: [asyncQuestionController],
  providers: [asyncQuestionService],
  imports: [NotificationModule, MailModule],
  exports: [asyncQuestionService],
})
export class asyncQuestionModule {}

@Module({
  controllers: [asyncQuestionController],
  providers: [asyncQuestionService],
  imports: [NotificationModule, MailTestingModule],
  exports: [asyncQuestionService],
})
export class asyncQuestionTestingModule {}
