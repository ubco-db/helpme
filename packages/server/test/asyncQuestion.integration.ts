import { AsyncQuestionModel } from 'asyncQuestion/asyncQuestion.entity';
import { CourseModel } from 'course/course.entity';
import { UserModel } from 'profile/user.entity';
import {
  UserFactory,
  CourseFactory,
  UserCourseFactory,
  AsyncQuestionFactory,
  VotesFactory,
} from './util/factories';
import { setupIntegrationTest } from './util/testUtils';
import { asyncQuestionModule } from 'asyncQuestion/asyncQuestion.module';
import { Role } from '@koh/common';

describe('AsyncQuestion Integration', () => {
  const supertest = setupIntegrationTest(asyncQuestionModule);
  let course: CourseModel;
  let TAuser: UserModel;
  let studentUser: UserModel;
  let studentUser2: UserModel;
  let asyncQuestion: AsyncQuestionModel;
  beforeEach(async () => {
    TAuser = await UserFactory.create({
      email: 'wskksw@student.ubc.ca',
      firstName: 'kevin',
      lastName: 'wang',
    });
    studentUser = await UserFactory.create({
      email: 'justino@ubc.ca',
      firstName: 'justin',
      lastName: 'oh',
    });
    studentUser2 = await UserFactory.create({
      email: 'tom@ubc.ca',
      firstName: 'tom',
      lastName: 'oh',
    });
    course = await CourseFactory.create({
      name: 'Test course',
    });
    await UserCourseFactory.create({
      user: TAuser,
      course,
      role: Role.TA,
    });
    await UserCourseFactory.create({
      user: studentUser,
      course,
      role: Role.STUDENT,
    });
    await UserCourseFactory.create({
      user: studentUser2,
      course,
      role: Role.STUDENT,
    });

    asyncQuestion = await AsyncQuestionFactory.create({ creator: studentUser });
  });
  describe('Async question creation', () => {
    it('Student can create a question', async () => {
      await supertest({ userId: studentUser.id })
        .post(`/asyncQuestions/${course.id}`)
        .send({
          questionAbstract: 'abstract',
          questionText: 'text',
        })
        .expect(201)
        .then((response) => {
          expect(response.body).toHaveProperty('status', 'AIAnswered');
          expect(response.body).toHaveProperty('closedAt', null);
          expect(response.body).toHaveProperty('questionText', 'text');
          expect(response.body.status).toBe('AIAnswered');
          expect(response.body.closedAt).toBeNull();
        });
    });
  });

  describe('Async question update', () => {
    it('Faculty can modify any question', async () => {
      await supertest({ userId: TAuser.id })
        .patch(`/asyncQuestions/${asyncQuestion.id}`)
        .send({
          questionAbstract: 'abstract',
          questionText: 'text1',
        })
        .expect(200)
        .then((response) => {
          expect(response.body).toHaveProperty('questionText', 'text1');
        });
      await supertest({ userId: TAuser.id })
        .patch(`/asyncQuestions/${asyncQuestion.id}`)
        .send({
          status: 'HumanAnswered',
        })
        .expect(200)
        .then((response) => {
          expect(response.body).toHaveProperty('status', 'HumanAnswered');
          expect(response.body).toHaveProperty('closedAt');
          expect(response.body.status).toBe('HumanAnswered');
          expect(response.body.closedAt).not.toBeNull();
        });
    });

    it('Student can modify their own question', async () => {
      await supertest({ userId: studentUser.id })
        .patch(`/asyncQuestions/${asyncQuestion.id}`)
        .send({
          status: 'HumanAnswered',
        })
        .expect(200)
        .then((response) => {
          expect(response.body).toHaveProperty('status', 'HumanAnswered');
          expect(response.body.status).toBe('HumanAnswered');
        });
    });

    it('Student cannot modify other students question', async () => {
      await supertest({ userId: studentUser2.id })
        .patch(`/asyncQuestions/${asyncQuestion.id}`)
        .send({
          status: 'HumanAnswered',
        })
        .expect(401);
    });
  });

  describe('POST /asyncQuestions/:qid/:vote', () => {
    it('should vote on a question', async () => {
      const student = await UserFactory.create();
      const question = await AsyncQuestionFactory.create({
        createdAt: new Date('2020-03-01T05:00:00.000Z'),
      });

      const response = await supertest({ userId: student.id }).post(
        `/asyncQuestions/${question.id}/1`,
      );

      expect(response.status).toBe(200);
      expect(response.body.vote).toBe(1);
    });

    it('should update an existing vote on a question', async () => {
      const student = await UserFactory.create();
      const question = await AsyncQuestionFactory.create({
        createdAt: new Date('2020-03-01T05:00:00.000Z'),
      });
      await VotesFactory.create({ userId: student.id, vote: 1, question });

      const response = await supertest({ userId: student.id }).post(
        `/asyncQuestions/${question.id}/-1`,
      );

      expect(response.status).toBe(200);
      expect(response.body.vote).toBe(0);
    });

    it('should not allow voting beyond the allowed range', async () => {
      const student = await UserFactory.create();
      const question = await AsyncQuestionFactory.create({
        createdAt: new Date('2020-03-01T05:00:00.000Z'),
      });
      await VotesFactory.create({ userId: student.id, vote: 1, question });

      const response = await supertest({ userId: student.id }).post(
        `/asyncQuestions/${question.id}/2`,
      );

      expect(response.status).toBe(200);
      expect(response.body.vote).toBe(1); // original vote
    });

    it('should not allow voting by unauthorized users', async () => {
      const question = await AsyncQuestionFactory.create({
        createdAt: new Date('2020-03-01T05:00:00.000Z'),
      });

      const response = await supertest().post(
        `/asyncQuestions/${question.id}/1`,
      );

      expect(response.status).toBe(401);
    });

    it('should return an error if the question does not exist', async () => {
      const student = await UserFactory.create();

      const response = await supertest({ userId: student.id }).post(
        `/asyncQuestions/9999/1`,
      );

      expect(response.status).toBe(404);
    });
  });
});
