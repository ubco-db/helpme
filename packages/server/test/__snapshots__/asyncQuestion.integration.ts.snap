// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncQuestion Integration GET /asyncQuestions/:courseId allows students to view their questions in their course 1`] = `
[
  {
    "aiAnswerText": "q1",
    "answerText": "answer",
    "authorSetVisible": false,
    "closedAt": null,
    "comments": [],
    "courseId": 1,
    "createdAt": "2025-01-01T00:00:00.000Z",
    "creator": {
      "anonId": 3,
      "colour": "#0097A7",
      "id": 2,
      "name": "justin oh",
      "photoURL": "https://example.com",
    },
    "creatorId": 2,
    "id": 1,
    "isAnonymous": true,
    "questionAbstract": "abstract",
    "questionText": "text",
    "questionTypes": [],
    "staffSetVisible": false,
    "status": "AIAnswered",
    "taHelpedId": null,
    "verified": false,
    "votes": [],
    "votesSum": 0,
  },
  {
    "aiAnswerText": "q2",
    "answerText": "answer",
    "authorSetVisible": false,
    "closedAt": null,
    "comments": [],
    "courseId": 1,
    "createdAt": "2025-01-01T00:00:00.000Z",
    "creator": {
      "anonId": 4,
      "colour": "#5A2E2E",
      "id": 2,
      "name": "justin oh",
      "photoURL": "https://example.com",
    },
    "creatorId": 2,
    "id": 2,
    "isAnonymous": true,
    "questionAbstract": "abstract",
    "questionText": "text",
    "questionTypes": [],
    "staffSetVisible": true,
    "status": "AIAnswered",
    "taHelpedId": null,
    "verified": false,
    "votes": [],
    "votesSum": 0,
  },
]
`;

exports[`AsyncQuestion Integration GET /asyncQuestions/:courseId does not show sensitive information for anonymous comments on questions unless they are the creator or staff 1`] = `
{
  "staff": [
    {
      "aiAnswerText": "q2",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [
        {
          "commentText": "comment1",
          "createdAt": "2025-01-02T00:00:00.000Z",
          "creator": {
            "anonId": 4,
            "colour": "#5A2E2E",
            "courseRole": "student",
            "id": 2,
            "isAuthor": true,
            "name": "justin oh",
            "photoURL": "https://example.com",
          },
          "id": 1,
          "isAnonymous": true,
          "questionId": 2,
        },
        {
          "commentText": "comment2",
          "createdAt": "2025-01-02T00:00:00.000Z",
          "creator": {
            "anonId": 5,
            "colour": "#0097A7",
            "courseRole": "student",
            "id": 3,
            "isAuthor": false,
            "name": "tom oh",
            "photoURL": "https://example.com",
          },
          "id": 2,
          "isAnonymous": true,
          "questionId": 2,
        },
      ],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 4,
        "colour": "#5A2E2E",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 2,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q3",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#5A2E2E",
        "id": 3,
        "name": "tom oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 3,
      "id": 3,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q1",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 3,
        "colour": "#0097A7",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 1,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
  ],
  "student": [
    {
      "aiAnswerText": "q2",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [
        {
          "commentText": "comment1",
          "createdAt": "2025-01-02T00:00:00.000Z",
          "creator": {
            "anonId": 4,
            "colour": "#5A2E2E",
            "courseRole": "student",
            "isAuthor": true,
            "photoURL": null,
          },
          "id": 1,
          "isAnonymous": true,
          "questionId": 2,
        },
        {
          "commentText": "comment2",
          "createdAt": "2025-01-02T00:00:00.000Z",
          "creator": {
            "anonId": 5,
            "colour": "#0097A7",
            "courseRole": "student",
            "id": 3,
            "isAuthor": false,
            "name": "tom oh",
            "photoURL": "https://example.com",
          },
          "id": 2,
          "isAnonymous": true,
          "questionId": 2,
        },
      ],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 4,
        "colour": "#5A2E2E",
        "name": "Anonymous",
        "photoURL": null,
      },
      "creatorId": 2,
      "id": 2,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q3",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#5A2E2E",
        "id": 3,
        "name": "tom oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 3,
      "id": 3,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
  ],
}
`;

exports[`AsyncQuestion Integration GET /asyncQuestions/:courseId does not show user information when question is anonymous except if the viewer is the creator or staff 1`] = `
{
  "author": [
    {
      "aiAnswerText": "q1",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 3,
        "colour": "#0097A7",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 1,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q9",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#1abc9c",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 4,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q2",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 4,
        "colour": "#5A2E2E",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 2,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
  ],
  "staff": [
    {
      "aiAnswerText": "q3",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#5A2E2E",
        "id": 3,
        "name": "tom oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 3,
      "id": 3,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q1",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 3,
        "colour": "#0097A7",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 1,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q9",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#1abc9c",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 4,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q2",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 4,
        "colour": "#5A2E2E",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 2,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
  ],
  "student": [
    {
      "aiAnswerText": "q3",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#5A2E2E",
        "id": 3,
        "name": "tom oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 3,
      "id": 3,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q9",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#1abc9c",
        "name": "Anonymous",
        "photoURL": null,
      },
      "creatorId": 2,
      "id": 4,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q2",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 4,
        "colour": "#5A2E2E",
        "name": "Anonymous",
        "photoURL": null,
      },
      "creatorId": 2,
      "id": 2,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
  ],
}
`;

exports[`AsyncQuestion Integration GET /asyncQuestions/:courseId prevents students from seeing questions without both author and staff setting them public with author approval required set true, but staff can see either way 1`] = `
{
  "staff": [
    {
      "aiAnswerText": "q3",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#5A2E2E",
        "id": 3,
        "name": "tom oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 3,
      "id": 3,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q1",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 3,
        "colour": "#0097A7",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 1,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q6",
      "answerText": "answer",
      "authorSetVisible": true,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#1abc9c",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 4,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q8",
      "answerText": "answer",
      "authorSetVisible": true,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 8,
        "colour": "#3498db",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 6,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q7",
      "answerText": "answer",
      "authorSetVisible": true,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 7,
        "colour": "#2ecc71",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 5,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q2",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 4,
        "colour": "#5A2E2E",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 2,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
  ],
  "student": [
    {
      "aiAnswerText": "q3",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#5A2E2E",
        "id": 3,
        "name": "tom oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 3,
      "id": 3,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q6",
      "answerText": "answer",
      "authorSetVisible": true,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#1abc9c",
        "name": "Anonymous",
        "photoURL": null,
      },
      "creatorId": 2,
      "id": 4,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
  ],
}
`;

exports[`AsyncQuestion Integration GET /asyncQuestions/:courseId prevents students from seeing questions without staff setting them public with author approval required set true, but staff can see either way 1`] = `
{
  "staff": [
    {
      "aiAnswerText": "q3",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#5A2E2E",
        "id": 3,
        "name": "tom oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 3,
      "id": 3,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q1",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 3,
        "colour": "#0097A7",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 1,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q6",
      "answerText": "answer",
      "authorSetVisible": true,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#1abc9c",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 4,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q8",
      "answerText": "answer",
      "authorSetVisible": true,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 8,
        "colour": "#3498db",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 6,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q7",
      "answerText": "answer",
      "authorSetVisible": true,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 7,
        "colour": "#2ecc71",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 5,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q2",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 4,
        "colour": "#5A2E2E",
        "id": 2,
        "name": "justin oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 2,
      "id": 2,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
  ],
  "student": [
    {
      "aiAnswerText": "q3",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#5A2E2E",
        "id": 3,
        "name": "tom oh",
        "photoURL": "https://example.com",
      },
      "creatorId": 3,
      "id": 3,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": false,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q6",
      "answerText": "answer",
      "authorSetVisible": true,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 6,
        "colour": "#1abc9c",
        "name": "Anonymous",
        "photoURL": null,
      },
      "creatorId": 2,
      "id": 4,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
    {
      "aiAnswerText": "q2",
      "answerText": "answer",
      "authorSetVisible": false,
      "closedAt": null,
      "comments": [],
      "courseId": 1,
      "createdAt": "2025-01-01T00:00:00.000Z",
      "creator": {
        "anonId": 4,
        "colour": "#5A2E2E",
        "name": "Anonymous",
        "photoURL": null,
      },
      "creatorId": 2,
      "id": 2,
      "isAnonymous": true,
      "questionAbstract": "abstract",
      "questionText": "text",
      "questionTypes": [],
      "staffSetVisible": true,
      "status": "AIAnswered",
      "taHelpedId": null,
      "verified": false,
      "votes": [],
      "votesSum": 0,
    },
  ],
}
`;
